src/cryptonote_core/service_node_list.cpp

```diff
@@ -2977,6 +2978,29 @@ namespace service_nodes
     return crypto::null_pkey;
   }
 
+  template<typename Iter, typename RandomGenerator>
+  Iter select_randomly(Iter start, Iter end, RandomGenerator& g) {
+      std::uniform_int_distribution<> dis(0, std::distance(start, end) - 1);
+      std::advance(start, dis(g));
+      return start;
+  }
+
+  template<typename Iter>
+  Iter select_randomly(Iter start, Iter end) {
+      static std::random_device rd;
+      static std::mt19937 gen(rd());
+      return select_randomly(start, end, gen);
+  }
+
+  crypto::public_key service_node_list::get_random_pubkey() {
+    //TODO sean
+    
+    //using service_nodes_infos_t = std::unordered_map<crypto::public_key, std::shared_ptr<const service_node_info>>;
+
```

Comment:
You need to hold a lock while accessing `service_nodes_infos` (to prevent other threads from modifying the service node list while you are accessing it):
```C++
std::lock_guard lock{m_sn_mutex};
```